pkgs/development/beam-modules/build-mix.nix

```diff
@@ -1,100 +1,71 @@
-{ stdenv, writeText, elixir, erlang, hex, lib }:
-
-{ name
-, version
-, src
-, setupHook ? null
-, buildInputs ? []
-, beamDeps ? []
-, postPatch ? ""
-, compilePorts ? false
-, installPhase ? null
-, buildPhase ? null
-, configurePhase ? null
-, meta ? {}
-, enableDebugInfo ? false
-, ... }@attrs:
-
-with lib;
+{ stdenv, lib, elixir, erlang, hex, rebar, rebar3, fetchMixDeps }:
 
-let
+{ name, version, src, depsPreConfigure ? null, nativeBuildInputs ? [ ]
+, meta ? { }, enableDebugInfo ? false, depsSha256, mixEnv ? "prod", ... }@attrs:
 
-  debugInfoFlag = lib.optionalString (enableDebugInfo || elixir.debugInfo) "--debug-info";
+let
 
-  shell = drv: stdenv.mkDerivation {
-          name = "interactive-shell-${drv.name}";
-          buildInputs = [ drv ];
+  shell = drv:
+    stdenv.mkDerivation {
+      name = "interactive-shell-${drv.name}";
+      buildInputs = [ drv ];
     };
 
-  bootstrapper = ./mix-bootstrap;
+  mixDeps = fetchMixDeps {
+    inherit src name mixEnv version;
+    preConfigure = depsPreConfigure;
+    sha256 = depsSha256;
+  };
 
-  pkg = self: stdenv.mkDerivation ( attrs // {
-    name = "${name}-${version}";
-    inherit version;
+  pkg = self:
+    stdenv.mkDerivation (attrs // {
+      name = "${name}-${version}";
+      inherit version;
 
-    dontStrip = true;
+      dontStrip = true;
 
-    inherit src;
+      inherit src;
 
-    setupHook = if setupHook == null
-    then writeText "setupHook.sh" ''
-       addToSearchPath ERL_LIBS "$1/lib/erlang/lib"
-    ''
-    else setupHook;
+      nativeBuildInputs = nativeBuildInputs ++ [ erlang hex elixir ];
 
-    inherit buildInputs;
-    propagatedBuildInputs = [ hex elixir ] ++ beamDeps;
+      MIX_ENV = mixEnv;
+      MIX_DEBUG = if enableDebugInfo then 1 else 0;
+      HEX_OFFLINE = 1;
+      DEBUG = if enableDebugInfo then 1 else 0; # for Rebar3 compilation
 
-    configurePhase = if configurePhase == null
-    then ''
-      runHook preConfigure
-      ${erlang}/bin/escript ${bootstrapper}
-      runHook postConfigure
-    ''
-    else configurePhase ;
+      postUnpack = ''
+        export HEX_HOME="$TMPDIR/hex"
+        export MIX_HOME="$TMPDIR/mix"
+        export MIX_DEPS_PATH="$TMPDIR/deps"
```

Comment:
Is there any particular reason to put mix deps in the directory, and not directly in `$src/deps`? I personally find that this makes the environment a bit different than usual, which can surprise users. At least it bite me a little bit while I was fixing the Mobilizon build, although this was definitely not a major issue.

Reply:
Can you detail a bit what you mean?
I have no problem with changing this, just want to understand where the issue lies.

Reply:
So regarding this, doing a cp on $src will not work
`cp: cannot create directory '/nix/store/ycqg5a39gzjcvim7cszk7q7260glw7lm-source/deps': Permission denied`
I learned something today as well.

I also tried not copying the dependency folder, but it doesn't work. The compilation of the dependencies will require that the dependency folder is writable. I also learned that today. I added a comment to prevent future people from being tricked.

Reply:
> Can you detail a bit what you mean?

When packaging mobilizon, I had to provide some locale files in a dependency directory, because it was downloading them at build time. My first instinct was to add this in pre-build:

```bash
cp -R ${cldrLocales}/* deps/ex_cldr/priv/cldr/locales/
```

That was because in a "normal" elixir project, the dependencies are under the same directory as the source code. But with the current implementation, this had to be corrected to:

```bash
cp -R ${cldrLocales}/* $MIX_DEPS_PATH/ex_cldr/priv/cldr/locales/
```

> So regarding this, doing a cp on $src will not work

What about if you do `cp --no-preserve=mode -R "${mixDeps}" ./deps`? I think the `$src` is the nix store path of the source, not the current build directory, my bad.

This really is just nitpicking, though, if it's a hassle, you can just forget about it ^^